Add pip check to py-bluepy

GianlucaFicarelli opened the pull request:
This is an example of a possible way to add a `pip check` command that will cause the installation to fail when it fails, if executed with
```
spack install --test=root py-bluepy
```
I tried it in the branch https://github.com/BlueBrain/spack/tree/faehigkeitenerweiterung/schlangeninstallationsueberpruefung
because this command always fails on the current develop branch (when installing `py-bluepy`, because of the dependency on `libflatindex`).
It should not be merged, but discussed:
- do we want to apply this check to other internal python packages?
- in that case, how to avoid to repeat the code on each `package.py` (changing the parent class, adding a configuration attribute, etc)

matz-e committed:
cherry-pick: add import_module checks for Python modules

Should allow us to automatically check more modules and alleviate the
need to create some private checks in different Jenkins CIs.

Original commit message:

PythonPackage: add import module smoke tests (#20023)

matz-e committed:
deploy: activate tests.

matz-e committed:
plumbing: call Mario!

matz-e committed:
deploy: log tests to XML

matz-e committed:
deploy: shuffle

matz-e committed:
py-ipyipyipy: add yet another dependency

matz-e committed:
py-pyzmq: add LD_LIBRARY_PATH for CFFI… :(

matz-e committed:
fix

matz-e committed:
fix da fix

matz-e committed:
fix da fix of da fix

GianlucaFicarelli committed:
Add pip check to bluepy

matz-e reviewed (changes requested):
(marking to request changes so that I don't overlook that label)

Thank you! This seems something we really want, and in a broader scope. I'm wondering if this could be added to `lib/spack/spack/build_systems/python.py` (or similar) to get it as a basic test for all Python packages. May require some work due to the added dependency, is the builtin pip shim not enough to have this run?

mgeplf commented:
That is really cool & so simple.  Upstreaming it would be worthwhile, but I can understand if upstream doesn't want to be tied to pip - it's a moving target that is constantly changing (although it's starting to slow down), and would probably create a pretty significant maintenance requirement.  @matz-e you're in contact w/ them, right? Perhaps you could float the idea?

matz-e commented:
> That is really cool & so simple. Upstreaming it would be worthwhile, but I can understand if upstream doesn't want to be tied to pip - it's a moving target that is constantly changing (although it's starting to slow down), and would probably create a pretty significant maintenance requirement. @matz-e you're in contact w/ them, right? Perhaps you could float the idea?

"contact" meaning I've filed PRs and tested some, then yes :)

We can always file PRs for upstream changes, but at times they move very slow, mostly due to the sheer volume of PRs.

matz-e commented on var/spack/repos/builtin/packages/py-bluepy/package.py:
```diff
@@ -80,3 +81,8 @@ def patch(self):
+    @run_after('install')
+    @on_package_attributes(run_tests=True)
+    def pip_check(self):
+        python('-mpip', 'check')
```
We also have this check here: https://github.com/BlueBrain/spack/blob/347d9a37e51fbd2de61d5aaf71c269159175119c/lib/spack/spack/build_systems/python.py#L356

Relies only on `setup.py` rather than `pip`… makes me wonder if we can find some way to hook that up automatically (and why that was not done before)

GianlucaFicarelli replied:
It seems to me that it would run `python setup.py check` that would check only metadata and restructuredtext, not dependencies, according to
https://github.com/pypa/setuptools/blob/c121d289da5d19cf6df2bf6b64ac28916a060161/setuptools/_distutils/command/check.py#L63-L77

matz-e replied:
Bummer.

matz-e committed:
py-pyzmq: attempt another fix

matz-e committed:
F 0MQ F F F

matz-e committed:
plotly: missing dependencies

matz-e committed:
flakey

matz-e committed:
plotly: more (kinda optional?) dependencies

GianlucaFicarelli commented:
> is the builtin pip shim not enough to have this run?

On BB5, I've seen that:
- the `pip` module is not available (unless loaded explicitly as in this PR)
- the `pip` command can be available or not, depending on the env (for example, when running spack inside a virtualenv, it uses that `pip` command)

So it seems more reliable if we load the module, but maybe there are other solutions? However the dependency is restricted to the type `test`, while it is not loaded for `run` or `build` without tests.

An additional note:
this test is run when installing the package using `spack install --test=root py-bluepy`(and skipped if the `--test` option is omitted) while the import module test (as in #1049) is run with `spack test run py-bluepy`, after the package has been installed.
I'm wondering if we want to run those tests in the same phase.

matz-e commented:
> > is the builtin pip shim not enough to have this run?
> 
> On BB5, I've seen that:
> 
>     * the `pip` module is not available (unless loaded explicitly as in this PR)
> 
>     * the `pip` command can be available or not, depending on the env (for example, when running spack inside a virtualenv, it uses that `pip` command)

I also forgot that Spack disables the basic built-in pip module. I would prefer not to load the module, the solution you have would work, since `pip` will get added to the PYTHONPATH. Need to ponder this more.
 
> So it seems more reliable if we load the module, but maybe there are other solutions? However the dependency is restricted to the type `test`, while it is not loaded for `run` or `build` without tests.
> 
> An additional note:
> this test is run when installing the package using `spack install --test=root py-bluepy`(and skipped if the `--test` option is omitted) while the import module test (as in #1049) is run with `spack test run py-bluepy`, after the package has been installed.
> I'm wondering if we want to run those tests in the same phase.

Yes, we should run them in the same phase, I'd actually prefer `spack test run` for this, to have a better installation test.

matz-e committed:
blargh

matz-e committed:
blubber

matz-e committed:
avoid unnecessary testing

in spack we trust!

matz-e committed:
refer to the right files

GianlucaFicarelli committed:
Run pip check in the package test method

and bump more-itertools to pass the test.

GianlucaFicarelli commented:
> Yes, we should run them in the same phase, I'd actually prefer `spack test run` for this, to have a better installation test.

To move forward, I moved `pip check` from the method decorated with `@run_after('install')` to the `test` method, that is executed with `spack test run`, instead of `spack install --test=root`.
I rebased on that [branch](https://github.com/BlueBrain/spack/tree/faehigkeitenerweiterung/schlangeninstallationsueberpruefung) to be able to try it.

> I also forgot that Spack disables the basic built-in pip module. I would prefer not to load the module, the solution you have would work, since `pip` will get added to the PYTHONPATH. Need to ponder this more.

I needed to change `depends_on('py-pip', type='test')` to `depends_on('py-pip', type='run')` because it seems that pip was not available. It looks strange to me, because according to the docs [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#install-tests): 
>> These tests run in an environment that provides access to the package and all of its dependencies, including `test`-type dependencies.

matz-e commented:
> I needed to change `depends_on('py-pip', type='test')` to `depends_on('py-pip', type='run')` because it seems that pip was not available. It looks strange to me, because according to the docs [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#install-tests):
> 
> > > These tests run in an environment that provides access to the package and all of its dependencies, including `test`-type dependencies.

Indeed, this is weird behavior. I'm kind of feeling like not too many people use these features. I tried hacking the parts of the code to include `test` dependencies, but haven't gotten far with that until now.

matz-e commented:
If I'm not mistaken, we've solved this differently for now (even though only for the Python modules deployed).
